langchain/callbacks/manager.py

```diff
@@ -0,0 +1,651 @@
+            run_id, self.handlers, self.inheritable_handlers, self.parent_run_id
+        )
+
+    async def on_chain_start(
+        self,
+        serialized: Dict[str, Any],
+        inputs: Dict[str, Any],
+        run_id: Optional[str] = None,
+        **kwargs: Any,
+    ) -> AsyncCallbackManagerForChainRun:
+        """Run when chain starts running."""
+        if run_id is None:
+            run_id = uuid.uuid4()
+
+        await _ahandle_event(
+            self.handlers,
+            "on_chain_start",
+            "ignore_chain",
+            serialized,
+            inputs,
+            run_id=run_id,
+            parent_run_id=self.parent_run_id,
+            **kwargs,
+        )
+
+        return AsyncCallbackManagerForChainRun(
+            run_id, self.handlers, self.inheritable_handlers, self.parent_run_id
+        )
+
+    async def on_tool_start(
+        self,
+        serialized: Dict[str, Any],
+        input_str: str,
+        run_id: Optional[str] = None,
+        parent_run_id: Optional[str] = None,
+        **kwargs: Any,
+    ) -> AsyncCallbackManagerForToolRun:
+        """Run when tool starts running."""
+        if run_id is None:
+            run_id = uuid.uuid4()
+
+        await _ahandle_event(
+            self.handlers,
+            "on_tool_start",
+            "ignore_agent",
+            serialized,
+            input_str,
+            run_id=run_id,
+            parent_run_id=parent_run_id,
+            **kwargs,
+        )
+
+        return AsyncCallbackManagerForToolRun(
+            run_id, self.handlers, self.inheritable_handlers, self.parent_run_id
+        )
+
+    @classmethod
+    def configure(
+        cls,
+        inheritable_callbacks: Optional[
+            Union[BaseCallbackManager, List[BaseCallbackHandler]]
+        ] = None,
+        local_callbacks: Optional[
+            Union[BaseCallbackManager, List[BaseCallbackHandler]]
+        ] = None,
+        verbose: bool = False,
+    ) -> Optional[BaseCallbackManager]:
+        """Configure the callback manager."""
+        return _configure(cls, inheritable_callbacks, local_callbacks, verbose)
+
+
+T = TypeVar("T", CallbackManager, AsyncCallbackManager)
+
+
+def _configure(
+    callback_manager_cls: Type[T],
+    inheritable_callbacks: Optional[Union[T, List[BaseCallbackHandler]]] = None,
+    local_callbacks: Optional[Union[T, List[BaseCallbackHandler]]] = None,
+    verbose: bool = False,
+) -> Optional[T]:
+    """Configure the callback manager."""
+    callback_manager: Optional[T] = None
+    if inheritable_callbacks or local_callbacks:
+        if isinstance(inheritable_callbacks, list) or not inheritable_callbacks:
+            callback_manager = callback_manager_cls(
+                handlers=inheritable_callbacks,
+                inheritable_handlers=inheritable_callbacks,
+            )
+        else:
+            callback_manager = inheritable_callbacks
+        callback_manager = copy.deepcopy(callback_manager)
```

Comment:
why call deep copy directly here rather than the copy method?

Reply:
Indeed, I ran into this. What's the logic behind a copy here anyway? I'd assume many callback managers—definitely mine—maintain state that cannot be copied such as file handles, network connections, asyncio.Futures etc.

I don't know this code that well, but why even copy at all? It seems somewhat antithetical to the rationale of a callback manager that everything makes a copy of it.